rust/src/utils.rs

```diff
@@ -496,3 +496,17 @@ unsafe impl<T: 'static + Clone> ToOCaml<DynBox<T>> for Abstract<T> {
 pub fn unwrap_abstract_vec<T>(v: Vec<Abstract<T>>) -> Vec<T> {
     v.into_iter().map(|Abstract(v)| v).collect()
 }
+
+pub fn maybe_release_runtime<T, F>(cr: &mut OCamlRuntime, should_release_runtime: bool, f: F) -> T
+where
+    F: FnOnce() -> T,
+{
+    if should_release_runtime {
```

Comment:
Is there a particular reason why we make releasing the runtime lock optional? I would've assumed that for these sorts of blocking operations it's always fine to release the runtime lock.

Reply:
Hmm yeah I wasn't too sure about this either. One failure mode that I could think of is if someone wraps the non-Async version in an In_thread.run (because it is in some larger function with other parallelizable bits), this function might unexpectedly run concurrently with other code. The user might end up with more concurrency than they'd like or maybe they have some logic to cache Data_frames to avoid duplicate work that gets ignored since the code no longer has the lock. 

What do you think? Am I overly cautious or off-base somewhere?

Reply:
> One failure mode that I could think of is if someone wraps the non-Async version in an In_thread.run (because it is in some larger function with other parallelizable bits), this function might unexpectedly run concurrently with other code.
Ah, I hadn't considered this case; that's a fair point. I see that upstream hasn't put a lot of thought into this problem (judging from https://github.com/pola-rs/polars/issues/8288), but at the same time if there's enough work to be done such that there's a ton of contention in Polar's thread pool, *not* releasing the runtime lock in an Async context to me seems like we'll end up in a worse situation (long async cycles, RPC timeouts, etc.).

> or maybe they have some logic to cache Data_frames to avoid duplicate work that gets ignored since the code no longer has the lock.
I'm not sure I follow; could you elaborate a bit on this case?

I think it might be worth leaving a comment in the code highlighting this possibility, but in the meantime it doesn't seem crazy to me to always release the runtime lock (or, create a lazy_static variable (or thread_local? I'm not sure how that would work with thread pools in async land) that toggles this behavior and defaults to always release the runtime lock). I'm guessing there will eventually be a lot of async functions exposed and if we need to pass a boolean variable around that seems super verbose and cumbersome, even if they are optional arguments.